### PR TITLE
Highlight features in histogram buckets on the map.

### PR DESCRIPTION
**Pull request by andreweland**
Highlight features in histogram buckets on the map, which involved moving histogram bucketing closer to the UI code. Draw the bucketed areas underneath buildings, although this will have to become feature type dependent in future.

### CODE REVIEW

**Comment by andreweland** [comment]
Example:
Screenshot 2023-10-04 at 11 30 04

File: src/diagonal.works/b6/api/functions/collections.go [hunk 1]
```diff
@@ -345,16 +345,16 @@ func flatten(_ *api.Context, c api.Collection) (api.Collection, error) {
 
 type bound struct {
```
Comment (xprogramari) [inline src/diagonal.works/b6/api/functions/collections.go:346]: can u not delete everything u re now doing in ui/histogram

File: src/diagonal.works/b6/api/functions/collections.go [hunk 2]
```diff
@@ -432,6 +432,10 @@ func uniform(kvs []kv) (buckets, error) {
 		}
 	}
 
+	for i := 0; i < len(b)/2; i++ {
+		b[i], b[len(b)-1-i] = b[len(b)-1-i], b[i]
```
Comment (xprogramari) [inline src/diagonal.works/b6/api/functions/collections.go:436]: why

File: src/diagonal.works/b6/api/collections.go [hunk 3]
```diff
@@ -599,7 +599,18 @@ func (a *ArrayAnyIntCollection) Next() (bool, error) {
 var _ Collection = &ArrayAnyIntCollection{}
 var _ Countable = &ArrayAnyIntCollection{}
 
-type HistogramCollection = ArrayAnyIntCollection
+type HistogramCollection struct {
+	Collection Collection
+	i          CollectionIterator
```
Comment (xprogramari) [inline src/diagonal.works/b6/api/collections.go:604]: y do u need the iterator separately ?

File: src/diagonal.works/b6/ui/histogram.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+		b[i], b[len(b)-1-i] = b[len(b)-1-i], b[i]
+	}
+
+	return b, nil
+}
+
+type kv struct {
+	key      interface{}
+	value    int            // key-count.
+	features []b6.FeatureID // features in this bucket, if relevant
```
Comment (xprogramari) [inline src/diagonal.works/b6/ui/histogram.go:183]: // features with this key, if identifiable ?

File: src/diagonal.works/b6/ui/histogram.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+		ok, err := i.Next()
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			break
+		}
+
+		var e *kv
+		if e, ok = m[i.Value()]; ok {
```
Comment (xprogramari) [inline src/diagonal.works/b6/ui/histogram.go:200]: u cant just ++ ?

File: src/diagonal.works/b6/ui/histogram.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+			e.value++
+		} else {
+			e = &kv{key: i.Value(), value: 1}
+			m[i.Value()] = e
+		}
+		if id, ok := i.Key().(b6.Identifiable); ok {
+			e.features = append(e.features, id.FeatureID())
```
Comment (xprogramari) [inline src/diagonal.works/b6/ui/histogram.go:207]: i m not sure i understand y we move the code actually, we could expose feature ids from upper layer also no? it feels a bit weird to have a histogram fn that just wraps the collection / but whatever

**Pull request merged**